src/Tempest/Http/src/Routing/Matching/GenericRouteMatcher.php

```diff
@@ -0,0 +1,86 @@
+<?php
+
+declare(strict_types=1);
+
+namespace Tempest\Http\Routing\Matching;
+
+use Psr\Http\Message\ServerRequestInterface as PsrRequest;
+use Tempest\Http\MatchedRoute;
+use Tempest\Http\Route;
+use Tempest\Http\RouteConfig;
+use Tempest\Http\Routing\Construction\MarkedRoute;
+
+final readonly class GenericRouteMatcher implements RouteMatcher
+{
+    public function __construct(private RouteConfig $routeConfig)
+    {
+    }
+
+    public function match(PsrRequest $request): ?MatchedRoute
+    {
+        // Try to match routes without any parameters
+        if (($staticRoute = $this->matchStaticRoute($request)) !== null) {
+            return $staticRoute;
+        }
+
+        // match dynamic routes
+        return $this->matchDynamicRoute($request);
+    }
+
+    private function matchStaticRoute(PsrRequest $request): ?MatchedRoute
+    {
+        $staticRoute = $this->routeConfig->staticRoutes[$request->getMethod()][$request->getUri()->getPath()] ?? null;
+
+        if ($staticRoute === null) {
+            return null;
+        }
+
+        return new MatchedRoute($staticRoute, []);
+    }
+
+    private function matchDynamicRoute(PsrRequest $request): ?MatchedRoute
+    {
+        // If there are no routes for the given request method, we immediately stop
+        $routesForMethod = $this->routeConfig->dynamicRoutes[$request->getMethod()] ?? null;
+        if ($routesForMethod === null) {
+            return null;
+        }
+
+        // Ensures matching regexes are available
+        $this->routeConfig->prepareMatchingRegexes();
```

Comment:
I'm wondering about caching at this point. We already have config cache, so maybe we should simply do this in the constructor of `RouteCache`, and store the compiled regex there?

Reply:
I do not see `RouteCache`, do you mean `RouteConfig` ? 

The reason why I do this: if you do not have discovery caching enabled. And you hit an endpoint, the RouteDiscovery will first build up the routingTree in the config. After which it needs to build up the regex. I do not (yet) have a hook to execute some logic after the Discovery phase. So I added this here to ensure it has.

If you do have an hook "after" the discovery phase. We could separate the whole route discovery  with tree building and the routing config. Which would be really amazing. But maybe a bit to much for this PR?

Reply:
That's a bit too much for this PR, indeed. Although, doesn't `KernelEvent::BOOTED` work for that? It's triggered after discovery: https://github.com/tempestphp/tempest-framework/blob/main/src/Tempest/Core/src/Kernel.php#L42

Reply:
Yeah it would. I did not expect it to be a thing. Nice! after this PR I will create one to separate the route construction from the route matching